docs/design/log-api.md

```diff
@@ -0,0 +1,148 @@
+
+This proposed design aims to:
+
+- be specification compliant,
+- have similar API to Trace and Metrics API,
+- take advantage of both OpenTelemetry and `slog` experience to achieve acceptable performance.
+
+### LoggerProvider
+
+The [`LoggerProvider` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#loggerprovider)
+is defined as an interface.
+
+```go
+type LoggerProvider interface{
+    Logger(name string, options ...LoggerOption) Logger
+}
+```
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface.
+
+```go
+type Logger interface{
+    Emit(ctx context.Context, options ...RecordOption)
+}
+```
+
+The `Logger` has `Emit(context.Context, options ...RecordOption` method.
+
+### Record
+
+The [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as a struct.
+
+```go
+type Record struct {
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          Severity
+	SeverityText      string
+	Body              string
+
+	// Allocation optimization: an inline array sized to hold
+	// the majority of log calls (based on examination of open-source
+	// code). It holds the start of the list of Attrs.
+	front [nAttrsInline]attribute.KeyValue
+
+	// The number of Attrs in front.
+	nFront int
+
+	// The list of Attrs except for those in front.
+	// Invariants:
+	//   - len(back) > 0 iff nFront == len(front)
+	//   - Unused array elements are zero. Used to detect mistakes.
+	back []attribute.KeyValue
+}
+
+const nAttrsInline = 5
+
+type Severity int
+
+const (
+	SeverityUndefined Severity = iota
+	SeverityTrace
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+	SeverityDebug
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+	SeverityInfo
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+	SeverityWarn
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+	SeverityError
+	SeverityError2
+	SeverityError3
+	SeverityError4
+	SeverityFatal
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+)
+```
+
+`Record` has `Attributes` and `AddAttributes` methods,
+like [`slog.Record.Attrs`](https://pkg.go.dev/log/slog#Record.Attrs)
+and [`slog.Record.AddAttrs`](https://pkg.go.dev/log/slog#Record.AddAttrs),
+in order to achieve high-performance when accessing and setting attributes efficiently.
+
+The `NewRecord(...RecordOption) Record` is a factory function used to create records using provided options.
+
+`Record` has a `Clone` method to allow copying records so that the SDK can offer concurrency safety.
+
+## Compatibility
+
+The backwards compatibility is achieved using the `embedded` design pattern
+that is already used in Trace API and Metrics API.
+
+## Benchmarking
+
+The benchmarks takes inspiration from [`slog`](https://pkg.go.dev/log/slog),
+because for the Go team it was also critical to create API that would be fast
+and interoperable with existing logging packages. [^1] [^2]
+
+## Rationale
```

Comment:
Were there any alternatives you considered and rejected?  e.g. Record as an interface, or anything else?

Reply:
Yes. Record as interface was one of them. Should I call it out? I would find it as a leaky abstraction and it would bring unnecessary overhead. For me the record is clearly a value object.

Reply:
I would definitely document anything you considered.  It helps people coming later to understand why it is the way it is.

Reply:
One argument against struct is that `Span` is an interface and not a struct and we aim for consistency. 

I understand the performance argument and agree it is important (not sure about "leaky abstraction"), but would want to see an _end-to-end_ benchmark showing how much difference a Record struct vs interface makes.

Reply:
@tigrannajaryan Do you have any other arguments why would it be good to model it as an interface? This is basically a value object without any behavior. Span is more complex as it actually has some behavior (like `End`). Moreover it was important for the span to encapsulate its the fields so that the user would not alter it during span lifetime. While log record is like "send and forget".

Reply:
Regarding API consistency the `Record` resembles more the instrument config structs (e.g. https://pkg.go.dev/go.opentelemetry.io/otel/metric#Float64CounterConfig) than a span.

Reply:
@pellared I don't have other arguments, and your arguments are pretty good. You may want to add them to the rationale :-)

Reply:
If nobody objects I will document this 👆  in Rationale section.

Reply:
Done